src/keystore/interfaces.ts

```diff
@@ -0,0 +1,89 @@
+import { messageApi, ciphertext, publicKey } from '@xmtp/proto'
+import { InvitationContext } from '../Invitation'
+import { ErrorCode } from './errors'
+
+type KeystoreError = {
+  error: string
+  code: ErrorCode
+}
+
+export type ResultOrError<T> = T | KeystoreError
+
+export type ConversationReference = {
+  topic: string
+  createdAt: Date
+  context?: InvitationContext
+}
+
+export type DecryptV1Request = {
+  payload: ciphertext.Ciphertext
+  peerKeys: publicKey.PublicKeyBundle
+  headerBytes: Uint8Array
+  isSender: boolean
+}
+
+export type DecryptV1Response = ResultOrError<{
+  decrypted: Uint8Array
+}>
+
+export type DecryptV2Request = {
+  payload: ciphertext.Ciphertext
+  headerBytes: Uint8Array
+  // Need to include contentTopic for the Keystore to know what topic key to use
+  contentTopic: string
+}
+
+export type DecryptV2Response = ResultOrError<{
+  decrypted: Uint8Array
+}>
+
+export type EncryptV1Request = {
+  recipient: publicKey.PublicKeyBundle
+  payload: Uint8Array
+  headerBytes: Uint8Array
+}
+
+export type EncryptResponse = ResultOrError<{
+  ciphertext: ciphertext.Ciphertext
+}>
+
+export type EncryptV2Request = {
+  contentTopic: string
+  message: Uint8Array
+  headerBytes: Uint8Array
+}
+
+export type CreateInviteRequest = {
+  recipient: publicKey.SignedPublicKeyBundle
+  createdAt: Date
+  context?: InvitationContext
+}
+
+export type CreateInviteResponse = {
+  conversation: ConversationReference
+  // The full bytes of the sealed invitation, which can then be published to the API
+  payload: Uint8Array
+}
+
+export interface Keystore {
+  // Decrypt a batch of V1 messages
+  decryptV1(req: DecryptV1Request[]): Promise<DecryptV1Response[]>
+  // Decrypt a batch of V2 messages
+  decryptV2(req: DecryptV2Request[]): Promise<DecryptV2Response[]>
+  // Encrypt a batch of V1 messages
+  encryptV1(req: EncryptV1Request[]): Promise<EncryptResponse[]>
+  // Encrypt a batch of V2 messages
+  encryptV2(req: EncryptV2Request[]): Promise<EncryptResponse[]>
+  // Decrypt and save a batch of invite for later use in decrypting messages on the invite topic
+  saveInvites(
+    req: messageApi.Envelope[]
```

Comment:
As I'm implementing a faux-Keystore interface in Rust WASM, I realize that `messageApi` is a proto for transport and is slightly more annoying to generate due to gRPC dependency. No need to change the API now but would it make sense to have the req be just message bytes? Seems like we pull envelope timestamp for a check, but don't really need it.